lib/sizzle.patched.js

```diff
@@ -1549,20 +1553,151 @@ function setFilters() {}
 setFilters.prototype = Expr.filters = Expr.pseudos;
 Expr.setFilters = new setFilters();
 
+/** Patch #6: Sorting tokens */
+
+/**
+ * Sorts the tokens in order to mitigate the performance issues caused by matching slow pseudos first:
+ * https://github.com/AdguardTeam/ExtendedCss/issues/55#issuecomment-364058745
+ * 
+ * @param {*} tokens a list of tokens to sort
+ */
+var TokenSorter = (function() {
+
+	/**
+	 * Splits compound selector into a list of simple selectors
+	 * 
+	 * @param {*} tokens Tokens to split into groups
+	 * @returns an array consisting of token groups (arrays) and relation tokens.
+	 */
+	var splitCompoundSelector = function(tokens) {
+
+		var groups = [];
+		var currentTokensGroup = [];
+		var maxIdx = tokens.length - 1;
+		for (var i = 0; i <= maxIdx; i++) {
+
+			var token = tokens[i];
+			var relative = Sizzle.selectors.relative[token.type];
+			if (relative) {
+				groups.push(currentTokensGroup);
+				groups.push(token);
+				currentTokensGroup = [];
+			} else {
+				currentTokensGroup.push(token);				
+			}
+
+			if (i === maxIdx) {
+				groups.push(currentTokensGroup);
+			}
+		}
+
+		return groups;
+	};
+
+	var TOKEN_TYPES_VALUES = {
+		// nth-child, etc, always go last
+		"CHILD": 100,
+		"ID": 90,
+		"CLASS": 80,
+		"TAG": 70,
+		"ATTR": 70,
+		"PSEUDO": 60
+	};
+
+	/** 
+	 * A function that defines the sort order.
+	 * Returns a value lesser than 0 if "left" is less than "right".
+	 */
+	var compareFunction = function(left, right) {
+		var leftValue = TOKEN_TYPES_VALUES[left.type];
+		var rightValue = TOKEN_TYPES_VALUES[right.type];
+		return leftValue - rightValue;
+	};
+
+	/**
+	 * Does all the sorting work
```

Comment:
Could you describe what the sorting work is about in more detail in the comment?

Reply:
TODO: Improve comment

Reply:
Done